File: js/ObjEncoder.js

/**
 * ObjEncoder serializes an object to an output stream.
 */
function afPickle_ObjEncoder(out, options) {
	this.out			= out;
	this.level			= 0;
	this.indent			= "\t";
	this.skipDefaults	= false;
	this.skipErrors		= false;
	this.curFieldType	= null;
	this.defaultObjs	= null;
	this.usings			= fan.sys.List.make(fan.sys.Str.$type, ["sys"]);
	if (options != null) this.initOptions(options);

	if (this.usings.size() > 0) {
		var that = this;
		this.usings.each(fan.sys.Func.make(
			fan.sys.List.make(fan.sys.Param.$type, [
				new fan.sys.Param("val","sys::Str", false),
			]),
			fan.sys.Void.$type,
			function(val) {
				that.w("using ").w(val).w("\n");
			})
		);
		this.w("\n");
	}
}

//////////////////////////////////////////////////////////////////////////
// Static
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.encode = function(obj) {
	var buf = fan.sys.StrBuf.make();
	var out = new fan.sys.StrBufOutStream(buf);
	new afPickle_ObjEncoder(out, null).writeObj(obj);
	return buf.toStr();
}

//////////////////////////////////////////////////////////////////////////
// Write
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.writeObj = function(obj) {
	if (obj == null) {
		this.w("null");
		return;
	}

	var t = typeof obj;
	if (t === "boolean") { this.w(obj.toString()); return; }
	if (t === "number")	{ this.w(obj.toString()); return; }
	if (t === "string")	{ this.wStrLiteral(obj.toString(), '"'); return; }

	var f = obj.$fanType;
	if (f === fan.sys.Float.$type)	 { fan.sys.Float.encode(obj, this); return; }
	if (f === fan.sys.Decimal.$type) { fan.sys.Decimal.encode(obj, this); return; }

	if (obj.$literalEncode) {
		obj.$literalEncode(this);
		return;
	}
	var type = fan.sys.ObjUtil.$typeof(obj);
	var ser = type.facet(fan.sys.Serializable.$type, false);
	if (ser != null) {
		if (ser.m_simple)
			this.writeSimple(type, obj);
		else
			this.writeComplex(type, obj, ser);
	}
	else
	{
		if (this.skipErrors) // NOTE: /* not playing nice in str - escape as unicode char */
			this.w("null /\u002A Not serializable: ").w(type.qname()).w(" */");
		else
			throw fan.sys.IOErr.make("Not serializable: " + type);
	}
}

//////////////////////////////////////////////////////////////////////////
// Simple
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.writeSimple = function(type, obj) {
	var str = fan.sys.ObjUtil.toStr(obj);
	this.wType(type).w('(').wStrLiteral(str, '"').w(')');
}

//////////////////////////////////////////////////////////////////////////
// Complex
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.writeComplex = function(type, obj, ser) {
	this.wType(type);

	var first = true;
	var defObj = null;
	if (this.skipDefaults) {
		var defType = fan.sys.ObjUtil.$typeof(obj).toNonNullable();
		if (this.defaultObjs[defType] != undefined)
			// use cached defObj if it exists
			defObj = this.defaultObjs[defType];
		else
		{
			// else attempt to instantiate default object for type,
			// this will fail if complex has it-block ctor
			try { defObj = fan.sys.ObjUtil.$typeof(obj).make(); } catch(e) {}
			this.defaultObjs[defType] = defObj;
		}
	}

	var fields = type.fields();
	for (var i=0; i<fields.size(); ++i) {
		var f = fields.get(i);

		// skip static, transient, and synthetic (once) fields
		if (f.isStatic() || f.isSynthetic() || f.hasFacet(fan.sys.Transient.$type))
			continue;

		// get the value
		var val = f.get(obj);

		// if skipping defaults
		if (defObj != null) {
			var defVal = f.get(defObj);
			if (fan.sys.ObjUtil.equals(val, defVal)) continue;
		}

		// if first then open braces
		if (first) { this.w(' ').w('{').w('\n'); this.level++; first = false; }

		// field name =
		this.wIndent().w(f.$name()).w('=');

		// field value
		this.curFieldType = f.type().toNonNullable();
		this.writeObj(val);
		this.curFieldType = null;

		this.w('\n');
	}

	// if collection
	if (ser.m_collection)
		first = this.writeCollectionItems(type, obj, first);

	// if we output fields, then close braces
	if (!first) { this.level--; this.wIndent().w('}'); }
}

//////////////////////////////////////////////////////////////////////////
// Collection (@collection)
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.writeCollectionItems = function(type, obj, first) {
	// lookup each method
	var m = type.method("each", false);
	if (m == null) throw fan.sys.IOErr.make("Missing " + type.qname() + ".each");

	// call each(it)
	var enc = this;
	var it	= fan.sys.Func.make(
		fan.sys.List.make(fan.sys.Param.$type),
		fan.sys.Void.$type,
		function(obj) {
			if (first) { enc.w('\n').wIndent().w('{').w('\n'); enc.level++; first = false; }
			enc.wIndent();
			enc.writeObj(obj);
			enc.w(',').w('\n');
			return null;
		});

	m.invoke(obj, fan.sys.List.make(fan.sys.Obj.$type, [it]));
	return first;
}

//////////////////////////////////////////////////////////////////////////
// List
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.writeList = function(list) {
	// get of type
	var of = list.of();

	// decide if we're going output as single or multi-line format
	var nl = this.isMultiLine(of);

	// figure out if we can use an inferred type
	var inferred = false;
	if (this.curFieldType != null && (this.curFieldType instanceof fan.sys.ListType)) {
		inferred = true;
	}

	// clear field type, so it doesn't get used for inference again
	this.curFieldType = null;

	// if we don't have an inferred type, then prefix of type
	if (!inferred) this.wType(of);

	// handle empty list
	var size = list.size();
	if (size == 0) { this.w("[,]"); return; }

	// items
	this.w('[');
	this.level++;
	for (var i=0; i<size; ++i) {
		if (i > 0) this.w(',');
		 if (nl) this.w('\n').wIndent();
		this.writeObj(list.get(i));
	}
	this.level--;
	if (nl) this.w('\n').wIndent();
	this.w(']');
}

//////////////////////////////////////////////////////////////////////////
// Map
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.writeMap = function(map) {
	// get k,v type
	var t = map.$typeof();

	// decide if we're going output as single or multi-line format
	var nl = this.isMultiLine(t.k) || this.isMultiLine(t.v);

	// figure out if we can use an inferred type
	var inferred = false;
	if (this.curFieldType != null && (this.curFieldType instanceof fan.sys.MapType)) {
		inferred = true;
	}

	// clear field type, so it doesn't get used for inference again
	this.curFieldType = null;

	// if we don't have an inferred type, then prefix of type
	if (!inferred) this.wType(t);

	// handle empty map
	if (map.isEmpty()) { this.w("[:]"); return; }

	// items
	this.level++;
	this.w('[');
	var first = true;
	var keys = map.keys();
	for (var i=0; i<keys.size(); i++) {
		if (first) first = false; else this.w(',');
		if (nl) this.w('\n').wIndent();
		var key = keys.get(i);
		var val = map.get(key);
		this.writeObj(key); this.w(':'); this.writeObj(val);
	}
	this.w(']');
	this.level--;
}

afPickle_ObjEncoder.prototype.isMultiLine = function(t) {
	return t.pod() != fan.sys.Pod.$sysPod;
}

//////////////////////////////////////////////////////////////////////////
// Output
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.wType = function(t) {
	return this.usings.contains(t.pod().m_name)
		? this.w(t.signature().split(t.pod().m_name + "::").join(""))
		: this.w(t.signature());
}

afPickle_ObjEncoder.prototype.wStrLiteral = function(s, quote) {
	var len = s.length;
	this.w(quote);
	// NOTE: these escape sequences are duplicated in FanStr.toCode()
	for (var i=0; i<len; ++i) {
		var c = s.charAt(i);
		switch (c) {
			case '\n': this.w('\\').w('n'); break;
			case '\r': this.w('\\').w('r'); break;
			case '\f': this.w('\\').w('f'); break;
			case '\t': this.w('\\').w('t'); break;
			case '\\': this.w('\\').w('\\'); break;
			case '"':	if (quote == '"') this.w('\\').w('"'); else this.w(c); break;
			case '`':	if (quote == '`') this.w('\\').w('`'); else this.w(c); break;
			case '$':	this.w('\\').w('$'); break;
			default:	 this.w(c);
		}
	}
	return this.w(quote);
}

afPickle_ObjEncoder.prototype.wIndent = function() {
	for (var i=0; i<this.level; ++i) this.w(this.indent);
	return this;
}

afPickle_ObjEncoder.prototype.w = function(s) {
	var len = s.length;
	for (var i=0; i<len; ++i)
		this.out.writeChar(s.charCodeAt(i));
	return this;
}

//////////////////////////////////////////////////////////////////////////
// Options
//////////////////////////////////////////////////////////////////////////

afPickle_ObjEncoder.prototype.initOptions = function(options) {
	this.indent			= options.get("indent",			this.indent);
	this.skipDefaults	= options.get("skipDefaults",	this.skipDefaults);
	this.skipErrors		= options.get("skipErrors",		this.skipErrors);

	if (typeof this.indent == "number")
		this.indent = " ".repeat(this.indent);
	else
		this.indent = this.indent.toString();

	if (this.skipDefaults)
		this.defaultObjs = {}

	if (options.containsKey("usings")) {
		this.usings = options.get("usings");
		if (this.usings == null)
			this.usings = fan.sys.List.make(fan.sys.Str.$type, []);
	}
}

File: java/fan/afPickle/ObjEncoder.java
package fan.afPickle;

import java.math.BigDecimal;
import java.util.HashMap;
import java.util.Iterator;
import java.util.Map.Entry;
import fan.sys.*;
import fanx.util.OpUtil;
import fanx.serial.Literal;

/**
 * ObjEncoder serializes an object to an output stream.
 */
public class ObjEncoder {

//////////////////////////////////////////////////////////////////////////
// Static
//////////////////////////////////////////////////////////////////////////

	public static String encode(Object obj, Map options) {
		StrBufOutStream out = new StrBufOutStream();
		new ObjEncoder(out, options).writeObj(obj);
		return out.string();
	}

//////////////////////////////////////////////////////////////////////////
// Constructor
//////////////////////////////////////////////////////////////////////////

	public ObjEncoder(OutStream out, Map options) {
		this.out = out;
		if (options != null) initOptions(options);
		else usings = List.make(Sys.StrType, new String[]{});
	
		if (usings.size() > 0) {
			for (Object using : usings.toArray()) {
				out.writeChars("using ").writeChars((String) using).writeChar('\n');
			}
			out.writeChar('\n');
		}
	}

//////////////////////////////////////////////////////////////////////////
// Write
//////////////////////////////////////////////////////////////////////////

	public void writeObj(Object obj) {
		if (obj == null) {
			w("null");
			return;
		}

		if (obj.getClass().getName().charAt(0) == 'j') {
			if (obj instanceof Boolean)		{ w(obj.toString());					return; }
			if (obj instanceof String)		{ wStrLiteral(obj.toString(), '"');		return; }
			if (obj instanceof Long)		{ w(obj.toString());					return; }
			if (obj instanceof Double)		{ encodeFloat((Double)obj, this);		return; }
			if (obj instanceof BigDecimal)	{ encodeDecimal((BigDecimal)obj, this);	return; }
		}

		// fanx.serial.Literal implementations
		if (obj instanceof Literal) {
			if (obj instanceof Duration)	{ w(((Duration)obj).toStr()); }
			if (obj instanceof List)		{ writeList((List) obj); }
			if (obj instanceof Map)			{ writeMap((Map) obj); }
			if (obj instanceof Slot)		{
				Slot slot	= (Slot) obj;
				Type parent	= slot.parent();
				w(parent.signature()).w("#").w(slot.name());
			}
			if (obj instanceof Type)		{ w(((Type)obj).signature()).w("#"); }
			if (obj instanceof Uri)			{ wStrLiteral(((Uri)obj).toStr(), '`'); }
			return;
		}

		Type type = FanObj.typeof(obj);
		Serializable ser = (Serializable)type.facet(Sys.SerializableType, false);
		if (ser != null) {
			if (ser.simple)
				writeSimple(type, obj);
			else
				writeComplex(type, obj, ser);
		}
		else {
			if (skipErrors)
				w("null /* Not serializable: ").w(type.qname()).w(" */");
			else
				throw IOErr.make("Not serializable: " + type);
		}
	}

	public static void encodeFloat(double self, ObjEncoder out) {
		if (Double.isNaN(self)) out.w("sys::Float(\"NaN\")");
		else if (self == Double.POSITIVE_INFINITY) out.w("sys::Float(\"INF\")");
		else if (self == Double.NEGATIVE_INFINITY) out.w("sys::Float(\"-INF\")");
		else out.w(Double.toString(self)).w("f");
	}
	
	public static void encodeDecimal(BigDecimal self, ObjEncoder out) {
		out.w(self.toString()).w("d");
	}
	
//////////////////////////////////////////////////////////////////////////
// Simple
//////////////////////////////////////////////////////////////////////////

	private void writeSimple(Type type, Object obj) {
		wType(type).w('(').wStrLiteral(FanObj.toStr(obj), '"').w(')');
	}

//////////////////////////////////////////////////////////////////////////
// Complex
//////////////////////////////////////////////////////////////////////////

	private void writeComplex(Type type, Object obj, Serializable ser) {
		wType(type);

		boolean first = true;
		Object defObj = null;
		if (skipDefaults) {		 
			Type defType = FanObj.typeof(obj).toNonNullable();
			if (defaultObjs.containsKey(defType))
				// use cached defObj if it exists
				defObj = defaultObjs.get(defType);
			else
			{
				// else attempt to instantiate default object for type,
				// this will fail if complex has it-block ctor
				try { defObj = defType.make(); } catch (Exception e) {}
				defaultObjs.put(defType, defObj);
			}
		}

		List fields = type.fields();
		for (int i=0; i<fields.sz(); ++i) {
			Field f = (Field)fields.get(i);

			// skip static, transient, and synthetic (once) fields
			if (f.isStatic() || f.isSynthetic() || f.hasFacet(Sys.TransientType))
				continue;

			// get the value
			Object val = f.get(obj);

			// if skipping defaults
			if (defObj != null) {
				Object defVal = f.get(defObj);
				if (OpUtil.compareEQ(val, defVal)) continue;
			}
			
			// if skipping nulls
			if (skipNulls && val == null)
				continue;

			// if first then open braces
			if (first) { w(' ').w('{').w('\n'); level++; first = false; }

			// field name =
			wIndent().w(f.name()).w('=');

			// field value
			curFieldType = f.type().toNonNullable();
			writeObj(val);
			curFieldType = null;

			w('\n');
		}

		// if collection
		if (ser.collection)
			first = writeCollectionItems(type, obj, first);

		// if we output fields, then close braces
		if (!first) { level--; wIndent().w('}'); }
	}

//////////////////////////////////////////////////////////////////////////
// Collection (@collection)
//////////////////////////////////////////////////////////////////////////

	private boolean writeCollectionItems(Type type, Object obj, boolean first) {
		// lookup each method
		Method m = type.method("each", false);
		if (m == null) throw IOErr.make("Missing " + type.qname() + ".each");

		// call each(it)
		EachIterator it = new EachIterator(first);
		m.invoke(obj, new Object[] { it });
		return it.first;
	}

	static final FuncType eachIteratorType = new FuncType(new Type[] { Sys.ObjType }, Sys.VoidType);

	class EachIterator extends Func.Indirect1 {
		EachIterator (boolean first) { super(eachIteratorType); this.first = first; }
		public Object call(Object obj) {
			if (first) { w('\n').wIndent().w('{').w('\n'); level++; first = false; }
			wIndent();
			writeObj(obj);
			w(',').w('\n');
			return null;
		}
		boolean first;
	}

//////////////////////////////////////////////////////////////////////////
// List
//////////////////////////////////////////////////////////////////////////

	public void writeList(List list) {
		// get of type
		Type of = list.of();

		// decide if we're going output as single or multi-line format
		boolean nl = isMultiLine(of);

		// figure out if we can use an inferred type
		boolean inferred = false;
		if (curFieldType != null && curFieldType.fits(Sys.ListType)) {
			inferred = true;
		}

		// clear field type, so it doesn't get used for inference again
		curFieldType = null;

		// if we don't have an inferred type, then prefix of type
		if (!inferred) wType(of);

		// handle empty list
		int size = list.sz();
		if (size == 0) { w("[,]"); return; }

		// items
		w('[');
		level++;
		for (int i=0; i<size; ++i) {
			if (i > 0) w(',');
			 if (nl) w('\n').wIndent();
			writeObj(list.get(i));
		}
		level--;
		if (nl) w('\n').wIndent();
		w(']');
	}

//////////////////////////////////////////////////////////////////////////
// Map
//////////////////////////////////////////////////////////////////////////

	public void writeMap(Map map) {
		// get k,v type
		MapType t = (MapType)map.typeof();

		// decide if we're going output as single or multi-line format
		boolean nl = isMultiLine(t.k) || isMultiLine(t.v);

		// figure out if we can use an inferred type
		boolean inferred = false;
		if (curFieldType != null && curFieldType.fits(Sys.MapType)) {
			inferred = true;
		}

		// clear field type, so it doesn't get used for inference again
		curFieldType = null;

		// if we don't have an inferred type, then prefix of type
		if (!inferred) wType(t);

		// handle empty map
		if (map.isEmpty()) { w("[:]"); return; }

		// items
		level++;
		w('[');
		boolean first = true;
		Iterator it = map.pairsIterator();
		while (it.hasNext()) {
			Entry e = (Entry)it.next();
			if (first) first = false; else w(',');
			if (nl) w('\n').wIndent();
			Object key = e.getKey();
			Object val = e.getValue();
			writeObj(key); w(':'); writeObj(val);
		}
		w(']');
		level--;
	}

	private boolean isMultiLine(Type t) {
		return t.pod() != Sys.sysPod;
	}

//////////////////////////////////////////////////////////////////////////
// Output
//////////////////////////////////////////////////////////////////////////

	public final ObjEncoder wType(Type t) {
		return usings.contains(t.pod().name())
			? w(t.signature().replace(t.pod().name() + "::", ""))
			: w(t.signature());
	}

	public final ObjEncoder wStrLiteral(String s, char quote) {
		int len = s.length();
		w(quote);
		// NOTE: these escape sequences are duplicated in FanStr.toCode()
		for (int i=0; i<len; ++i) {
			char c = s.charAt(i);
			switch (c) {
				case '\n': w('\\').w('n'); break;
				case '\r': w('\\').w('r'); break;
				case '\f': w('\\').w('f'); break;
				case '\t': w('\\').w('t'); break;
				case '\\': w('\\').w('\\'); break;
				case '"':	if (quote == '"') w('\\').w('"'); else w(c); break;
				case '`':	if (quote == '`') w('\\').w('`'); else w(c); break;
				case '$':	w('\\').w('$'); break;
				default:	 w(c);
			}
		}
		return w(quote);
	}

	public final ObjEncoder wIndent() {
		for (int i=0; i<level; ++i) out.writeChars(indent);
		return this;
	}

	public final ObjEncoder w(String s) {
		int len = s.length();
		for (int i=0; i<len; ++i)
			out.writeChar(s.charAt(i));
		return this;
	}

	public final ObjEncoder w(char ch) {
		out.writeChar(ch);
		return this;
	}

//////////////////////////////////////////////////////////////////////////
// Options
//////////////////////////////////////////////////////////////////////////

	private void initOptions(Map options) {
		indent			= option(options, "indent", indent);
		skipDefaults	= (boolean) options.get("skipDefaults",	skipDefaults);
		skipErrors		= (boolean) options.get("skipErrors",	skipErrors);
		skipNulls		= (boolean) options.get("skipNulls",	skipNulls);
		usings			= (List)	options.get("usings",		usings);
		if (skipDefaults)
			defaultObjs = new HashMap();
		if (usings == null) {
			// "usings" is legacy, prefer "using" instead
			Object using = (Object)	options.get("using",		usings);
			if (using instanceof List)
				usings = (List) using;
			else
			if (using != null && FanStr.trimToNull(using.toString()) != null)
				usings = FanStr.splitws(using.toString());
		}
		if (usings == null)
			usings		= List.make(Sys.StrType, new String[]{});
	}

	private static String option(Map options, String name, String def) {
		Object val = (Object) options.get(name);
		if (val == null) return def;
		
		if (val instanceof Long)
			return String.format("%" + ((Long) val) + "s", " ");
		
		return val.toString();
	}

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

	OutStream	out;
	int			level				= 0;
	String		indent			= "\t";
	boolean		skipDefaults	= false;
	boolean		skipErrors		= false;
	boolean		skipNulls		= false;
	Type		curFieldType;
	HashMap		defaultObjs;
	List		usings;
}
